runtime/src/main/java/org/corfudb/runtime/collections/SMRMultiIndex.java

```diff
@@ -0,0 +1,374 @@
+package org.corfudb.runtime.collections;
+
+
+import lombok.AllArgsConstructor;
+import lombok.Data;
+import lombok.EqualsAndHashCode;
+import org.corfudb.annotations.*;
+
+import javax.annotation.Nullable;
+import java.util.*;
+
+/**
+ * //FIXME: These indexes are secondary indexes on a data structure. These should be more like a library that
+ * // FIXME: extends the functionality of SMRMap and does not need to be a separate standalone datastructure.
+ * This object implements a multi-indexed map.
+ *
+ * Maps are indexed by their row index, R, as well as multiple Named indexes
+ * calculated at insert time using index functions, which generate
+ * indexes from the inserted value.
+ *
+ * The Main Indexing Data Structure.
+ * Each named index maintains a <Code>Map</Code> of indexKeys that in turn
+ * point to a <Code>Map</Code> of keys and values from the main <Code>Map</Code>
+ * datastructure. This multi-level <Code>Map</Code> provides flexibility for the current use case.
+ *
+ *  ____________________________________________________________________________________________
+ * | <String>IndexName</String> -->  ___________________________________________________________|
+ * |                                 | <I>IndexKey1</I> --> ____________________________________|
+ * |                                 |                     |<K>PrimaryKey1</K> --><V>Value1</V> |
+ * |                                 |                     |<K>PrimaryKey2</K> --><V>Value2</V> |
+ * |                                 |                     |____________________________________|
+ * |                                 |                     |____________________________________|
+ * |                                 | <I>IndexKey2</I> -->|<K>PrimaryKey1</K> --><V>Value1</V> |
+ * |                                 |                     |<K>PrimaryKey3</K> --><V>Value3</V> |
+ * |_________________________________|_____________________|____________________________________|
+ *
+ *
+ *
+ * Created by mwei on 4/7/17.
+ */
+//FIXME: having four generics is clunky. We need to come up with a simpler interface.
+@CorfuObject
+public class SMRMultiIndex<K, V, I> implements Map<K, V> {
+
+    //TODO Index function is too generic and can be source of inefficiency
+    //TODO Maybe there should be efficient implementations that can do efficient
+    //TODO index generation.
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this TODO comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=squid%3AS1135)